File: Binary_Tree_Maximum_Path_Sum.cpp
/*124. Binary Tree Maximum Path Sum Add to List
Description  Submission  Solutions
Total Accepted: 88554
Total Submissions: 349937
Difficulty: Hard
Contributors: Admin
Given a binary tree, find the maximum path sum.

For this problem, a path is defined as any sequence of nodes from some starting
node to any node in the tree along the parent-child connections. The path must
contain at least one node and does not need to go through the root.

For example:
Given the below binary tree,

       1
      / \
     2   3
Return 6.

Hide Company Tags Microsoft Baidu
Hide Tags Tree Depth-first Search
Hide Similar Problems (E) Path Sum (M) Sum Root to Leaf Numbers
*/

/*A recursive method maxPath(TreeNode node)
(1) computes the maximum path sum with highest node is the input node, update
maximum if necessary (2) returns the maximum sum of the path that can be
extended to input node's parent.*/

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
 private:
  int maxValue;

 public:
  int maxPathSum(TreeNode* root) {
    maxValue = INT_MIN;  // return the minvalue of the integer
    maxPath(root);
    return maxValue;
  }

  int maxPath(TreeNode* root) {
    if (!root) return 0;
    int left = max(0, maxPath(root->left));
    int right = max(0, maxPath(root->right));
    maxValue = max(maxValue, left + right + root->val);
    return max(left, right) + root->val;
  }
};

File: Binary_Tree_Maximum_Path_Sum.java
/*
Given a non-empty binary tree, find the maximum path sum.

For this problem, a path is defined as any sequence of nodes from some starting node to any node in the tree along the parent-child connections. The path must contain at least one node and does not need to go through the root.

Example 1:

Input: [1,2,3]

       1
      / \
     2   3

Output: 6
Example 2:

Input: [-10,9,20,null,null,15,7]

   -10
   / \
  9  20
    /  \
   15   7

Output: 42

*/

/*
Solution 1: recursive
Initiate max_sum as the smallest possible integer and call max_gain(node = root).
Implement max_gain(node) with a check to continue the old path/to start a new path:
Base case : if node is null, the max gain is 0.
Call max_gain recursively for the node children to compute max gain from the left and right subtrees : left_gain = max(max_gain(node.left), 0) and
right_gain = max(max_gain(node.right), 0).
Now check to continue the old path or to start a new path. To start a new path would cost price_newpath = node.val + left_gain + right_gain. Update max_sum if it's better to start a new path.
For the recursion return the max gain the node and one/zero of its subtrees could add to the current path : node.val + max(left_gain, right_gain).
//
Time complexity : O(N) where N is number of nodes, since we visit each node not more than 2 times.
Space complexity : \mathcal{O}(\log(N))O(log(N)). We have to keep a recursion stack of the size of the tree height, which is \mathcal{O}(\log(N))O(log(N)) for the binary tree.
*/


/*A recursive method maxPath(TreeNode node)
(1) computes the maximum path sum with highest node is the input node, update maximum if necessary (2) returns the maximum sum of the path that can be extended to input node's parent.*/
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    private int maxValue;

    public int maxPathSum(TreeNode root) {
        maxValue = Integer.MIN_VALUE;
        maxPath(root);
        return maxValue;
    }

    public int maxPath(TreeNode root){
        if(root == null) return 0;
        int left = Math.max(0, maxPath(root.left));
        int right = Math.max(0, maxPath(root.right));
        maxValue = Math.max(maxValue, left + right + root.val);
        return Math.max(left, right) + root.val;
    }
}
